### PR TITLE
Updated security guidelines

### PR DESCRIPTION
**Pull request by buskamuza**
- added the Security section to older Magento versions
- added more restrictions

## This PR is a:

- [ ] New topic
- [x] Content update
- [x] Content fix or rewrite
- [ ] Bug fix or improvement

## Summary

When this pull request is merged, it will:
- add more restrictions to the Security part of [Technical Guidelines](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- add Security section to older Magento versions (it was missing there)

## Additional information

- https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.1/coding-standards/technical-guidelines.html
- https://devdocs.magento.com/guides/v2.0/coding-standards/technical-guidelines.html

whatsnew
TBD when ready to merge

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/devdocs?pullRequest=2886) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/devdocs?pullRequest=2886) before we can accept your contribution.<br/><hr/>**Olga Kopylova** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by buskamuza** [comment]
Not ready for merge yet.

**Comment by lorikrell** [comment]
@buskamuza - Happy to review and help merge when this is ready.

**Comment by magento-cicd2** [comment]
Can one of the admins verify this patch?

**Comment by buskamuza** [comment]
This can be published (reviewed by @piotrekkaminski ).

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 1]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.2.6. Limit or increasingly delay failed login attempts. Log all failures and alert administrators when credential stuffing, brute force, or other attacks are detected.
+
+15.2.7. Use a server-side, secure, built-in session manager that generates a new random session ID with high entropy after login. Session IDs should not be in the URL, be securely stored and invalidated after logout, idle, and absolute timeouts.
+
+15.3. Cross-Site Scripting (XSS) protection.
+
+15.3.1. Sanitize input; escape output.
+
+15.3.2. Follow [templates XSS security guidelines](https://devdocs.magento.com/guides/v2.2/frontend-dev-guide/templates/template-security.html) for escaping output.
```
Comment (lorikrell) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
15.3.2. Follow [templates XSS security guidelines]({{ page.baseurl }}/frontend-dev-guide/templates/template-security.html) for escaping output.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 2]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.2.6. Limit or increasingly delay failed login attempts. Log all failures and alert administrators when credential stuffing, brute force, or other attacks are detected.
+
+15.2.7. Use a server-side, secure, built-in session manager that generates a new random session ID with high entropy after login. Session IDs should not be in the URL, be securely stored and invalidated after logout, idle, and absolute timeouts.
+
+15.3. Cross-Site Scripting (XSS) protection.
+
+15.3.1. Sanitize input; escape output.
+
+15.3.2. Follow [templates XSS security guidelines](https://devdocs.magento.com/guides/v2.2/frontend-dev-guide/templates/template-security.html) for escaping output.
```
Comment (lorikrell) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
15.3.2. Follow [templates XSS security guidelines]({{ page.baseurl }}/frontend-dev-guide/templates/template-security.html) for escaping output.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 3]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.10. Remote Code Execution (RCE) protection.
+
+15.10.1. `eval()`, `passthru()`, `system()`, `shell_exec()`, `serialize()`, `unserialize()`, `md5()`, `srand()`, `mt_srand()` SHOULD NOT be used.
+
+15.10.2. User-submitted values SHOULD NOT be passed directly to `include*()`, `require*()`, `create_function()`, `fopen()`, `preg_replace()`.
+
+15.10.3. Variable functions SHOULD NOT be used if the variable values are submitted by the user.
+
+15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by entire application in a centralize manner.
```
Comment (lorikrell) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by the entire application in a centralize manner.
```

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 4]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.10. Remote Code Execution (RCE) protection.
+
+15.10.1. `eval()`, `passthru()`, `system()`, `shell_exec()`, `serialize()`, `unserialize()`, `md5()`, `srand()`, `mt_srand()` SHOULD NOT be used.
+
+15.10.2. User-submitted values SHOULD NOT be passed directly to `include*()`, `require*()`, `create_function()`, `fopen()`, `preg_replace()`.
+
+15.10.3. Variable functions SHOULD NOT be used if the variable values are submitted by the user.
+
+15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by entire application in a centralize manner.
```
Comment (lorikrell) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by the entire application in a centralize manner.
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 5]
```diff
@@ -759,17 +759,21 @@ class SampleEventObserverThatModifiesInputs
...
-15.10.1. SHOULD NOT use eval(), passthru(), system(), shell_exec(), serialize(), unserialize(), md5(), srand(), mt_srand()
+15.10.1. `eval()`, `passthru()`, `system()`, `shell_exec()`, `serialize()`, `unserialize()`, `md5()`, `srand()`, `mt_srand()` SHOULD NOT be used.
 
-15.10.2. SHOULD NOT directly pass user-submitted values to include*(), require*(), create_function(), fopen(), preg_replace()
+15.10.2. User-submitted values SHOULD NOT be passed directly to `include*()`, `require*()`, `create_function()`, `fopen()`, `preg_replace()`.
 
-15.10.3. SHOULD NOT use variable functions if the variable values are submitted by the user.
+15.10.3. Variable functions SHOULD NOT be used if the variable values are submitted by the user.
+
+15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by entire application in a centralize manner.
```
Comment (lorikrell) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
15.11. Security capabilities SHOULD be implemented either on the Magento Framework level or in a dedicated module(s) and utilized by the entire application in a centralize manner.
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 6]
```diff
@@ -759,17 +759,21 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`) except files that are intended to be publicly accessible.
```
Comment (lorikrell) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`), except files that are intended to be publicly accessible.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 7]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`) except files that are intended to be publicly accessible.
```
Comment (lorikrell) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`), except files that are intended to be publicly accessible.
```

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 8]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`) except files that are intended to be publicly accessible.
```
Comment (lorikrell) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
15.12. Files MUST be secured by a web server configuration (e.g., `.htaccess` or `nginx.conf`), except files that are intended to be publicly accessible.
```

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 9]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.5.2. A segmented application architecture that provides effective, secure separation between components or tenants, with segmentation, containerization, or cloud security groups (ACLs).
+
+15.5.3. Sending security directives to clients, e.g. [Security Headers](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project).
+
+15.6. Sensitive Data Exposure protection.
+
+15.6.1. Exceptions/Notices/Warnings should be caught and logged.
+
+15.6.2. No error output should be displayed to user, some standard message should appear instead.
```
Comment (lorikrell) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
15.6.2. Error output should not be displayed to the user. Display standard messages to inform the user.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 10]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.5.2. A segmented application architecture that provides effective, secure separation between components or tenants, with segmentation, containerization, or cloud security groups (ACLs).
+
+15.5.3. Sending security directives to clients, e.g. [Security Headers](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project).
+
+15.6. Sensitive Data Exposure protection.
+
+15.6.1. Exceptions/Notices/Warnings should be caught and logged.
+
+15.6.2. No error output should be displayed to user, some standard message should appear instead.
```
Comment (lorikrell) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
15.6.2. Error output should not be displayed to the user. Display standard messages to inform the user.
```

File: guides/v2.1/coding-standards/technical-guidelines.md [hunk 11]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.6.3. Logs should not be excessive, e.g. PDO exception contains MySQL credentials that should not be logged.
+
+15.7. Cross-Site Request Forgery (CSRF) protection.
+
+15.7.1. CSRF tokens mechanism should be utilized.
+
+15.7.2. All data manipulation requests should be made with POST requests.
+
+15.8. Frequently update the third-party libraries used in the project/component to eliminate known vulnerabilities.
```
Comment (lorikrell) [inline guides/v2.1/coding-standards/technical-guidelines.md]:
```suggestion
15.8. Frequently update third-party libraries used in the project/component to eliminate known vulnerabilities.
```

File: guides/v2.0/coding-standards/technical-guidelines.md [hunk 12]
```diff
@@ -695,6 +695,80 @@ class SampleEventObserverThatModifiesInputs
...
+
+15.6.3. Logs should not be excessive, e.g. PDO exception contains MySQL credentials that should not be logged.
+
+15.7. Cross-Site Request Forgery (CSRF) protection.
+
+15.7.1. CSRF tokens mechanism should be utilized.
+
+15.7.2. All data manipulation requests should be made with POST requests.
+
+15.8. Frequently update the third-party libraries used in the project/component to eliminate known vulnerabilities.
```
Comment (lorikrell) [inline guides/v2.0/coding-standards/technical-guidelines.md]:
```suggestion
15.8. Frequently update third-party libraries used in the project/component to eliminate known vulnerabilities.
```

**Comment by lorikrell** [comment]
run tests

**Comment by dshevtsov** [comment]
@lorikrell please update whatsnew

**Comment by lorikrell** [comment]
@dshevtsov Updated!

**Pull request merged**